### DISCUSSION TITLE
GitLog() for sub directory

### DISCUSSION THREAD

**Discussion by snarum** [cake-build/cake#3107] [Extension Q&A] [answered]
I can't make GitLog alias work for a subdirectory in my repository. from git command line I would have written this:
```
git log -- c:\repos\myrepo\mysubfolder
```

Anybody know a way to do this? Or if it's supported at all?

**Comment by nils-a** [comment]
At a quick glance I would say it is currently not possible using `GitLog()`.
The alternative I see at the moment is something like 

```cs
StartProcess("git.exe", new ProcessSettings {
  Arguments = "log -- mysubfolder",
  WorkingDirectory = "c:\\repos\\myrepo"
});
```

Obviously this would only be a crude workaround, so I would suggest you raise a ticket with [Cake_Git](https://github.com/cake-contrib/Cake_Git). (Of course PRs are always welcome 😉 )

**Reply by snarum** [reply]
Thanks! I was just going down the StartProcess now :)

**Comment by pascalberger** [comment, marked as answer]
Cake.Git is basically just a wrapper around LibGit2Sharp.  As a workaround if there's currently no alias available you can use LibGit2Sharp directly, after adding the Cake.Git addin:

```csharp
using (var repo = new Repository(@"c:\repos\myrepo"))
{
    IEnumerable<LogEntry> history = repo.Commits.QueryBy(@"mysubfolder");
}
```

As a second step you can provide a PR to Cake.Git to add an alias around the specific functionality, and once this is merged and released, you can remove the direct calls to LibGit2Sharp in your script, with the alias from Cake.Git.

**Reply by snarum** [reply]
Excellent!

Do you also know if it is possible with LibGit2Sharp to filter the commit log with both path and tag? The QueryBy() has two overloads, one with path and one with Filter, but I can't find a way to combine the two.

Similar from command line:

```
git log mytag -- myfolder
```

**Reply by pascalberger** [reply]
I never did this myself, but looking through source code [QueryBy(string path, CommitFilter filter)](https://github.com/libgit2/libgit2sharp/blob/7fc4be5193dbdd08538b4b150332b5a73770e0f6/LibGit2Sharp/IQueryableCommitLog.cs#L31) seems to be what you're looking for:

```csharp
/// <summary>
/// Returns the list of commits of the repository representing the history of a file beyond renames.
/// </summary>
/// <param name="path">The file's path.</param>
/// <param name="filter">The options used to control which commits will be returned.</param>
/// <returns>A list of file history entries, ready to be enumerated.</returns>
IEnumerable<LogEntry> QueryBy(string path, CommitFilter filter);
```
